File: lua/grpc_connection.lua
-------------------------------------------------------------------------------
-- this module is just used for server,
-- the features for client is not implemented
-------------------------------------------------------------------------------
local libpickle = require("libpickle")
local socket = require("socket")
local config = require("grpc_config")
local Connection = {}

local __simple_types = {'nil', 'boolean', 'number', 'string'}
local simple_types = {}
for k, v in pairs(__simple_types) do
    simple_types[v] = true
end

local weak_value_metatbl = {}
weak_value_metatbl.__mode = "v"

function Connection:new(o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    o.connected = false
    o.__sock = nil
    o.__local_objects = {}
    o.__oid = 1
    setmetatable(o.__local_objects, weak_value_metatbl)
    return o
end

function Connection:accept(server_sock)
    if not self.connected then
        server_sock:settimeout(0)   -- in seconds
        self.__sock = server_sock:accept()
        if self.__sock ~= nil then
            self.connected = true
            self.__sock:settimeout(0)
            return self.__sock:getpeername()
        end
    end
end

function Connection:shutdown()
    if self.connected then
        self:send_shutdown()
        self.__sock:shutdown("both")
        self.__sock:close()
        self.connected = false
    end
end

function Connection:send_reply(seq_num, action_type, data)
    local boxed_data = self:__box_reply(data)
    return self:__send(config.msg.reply, seq_num, action_type, boxed_data)
end

function Connection:send_exception(seq_num, data)
    self:__send(config.msg.exception, seq_num, 0, data)
    return seq_num
end

function Connection:send_shutdown()
    local status, res = pcall(self.__send, self, config.msg.shutdown, 0, 0, 0)
    return res
end

function Connection:__send(msg_type, seq_num, action_type, boxed_data)
    if self.connected == false then
        return -1
    end
    local pickled_data = libpickle.dump({msg_type, seq_num, action_type, boxed_data})
    local data_size = string.len(pickled_data)
    if data_size > 99999999 then
        return -1, 'data size is too large'
    end
    self.__sock:send(string.format("%08d", data_size) .. pickled_data)
    return seq_num
end

function Connection:__box_reply(obj)
    if simple_types[type(obj)] then
        return {config.label.value, obj}
    else
        --[[
        local cache_item = {}
        setmetatable(cache_item, weak_value_metatbl)
        cache_item[self.__oid] = obj
        ]]
        self.__local_objects[self.__oid] = obj
        self.__oid = self.__oid + 1
        return {config.label.remote_ref, {self.__oid-1, tostring(obj), tostring(obj)}}
    end
end

function Connection:recv(timeout)
    if not self.connected then
        return nil
    end
    timeout = timeout or -1
    local recvt, sendt, status = socket.select({self.__sock}, nil, timeout)
    if status == "timeout" then
        return nil
    end
    local pickled_data = ""
    if #recvt then
        local data_size, status = self.__sock:receive(8)
        if status == 'closed' then
            self:shutdown()
            print 'shutdown'
            return config.msg.shutdown
        end
        if status == 'timeout' and data == nil then
            print 'data size timeout'
            return nil
        end
        data_size = tonumber(data_size)
        pickled_data, status = self.__sock:receive(data_size)
        if pickled_data == nil then
            print 'pickle data is nil'
            return nil
        end
    end

    local package = libpickle.load(pickled_data)
    local msg_type = package[1]
    local seq_num = package[2]
    local action_type = package[3]
    local data = package[4]

    if msg_type == config.msg.request then
        local status, unboxed_data = pcall(self.__unbox, self, data)
        if status == false then
            local oid = string.gmatch(unboxed_data, "oid=(%w+)")()
            local error_msg = 'object oid=' .. oid .. ' has been deleted'
            self:send_exception(seq_num, error_msg)
            return nil
        else
            return msg_type, seq_num, action_type, unboxed_data
        end
    elseif msg_type == config.msg.shutdown then
        self:shutdown()
        return nil
    end
end

function Connection:__unbox(package)
    local label = package[1]
    local value = package[2]
    if label == config.label.value then
        return value
    elseif label == config.label.tuple or label == config.label.list then
        local res = {}
        for i, v in ipairs(value) do
            res[i] = self:__unbox(v)
        end
        return res
    elseif label == config.label.dict then
        local res = {}
        for i in pairs(value) do
            res[self:__unbox(i)] = self:__unbox(value[i])
        end
        return res
    elseif label == config.label.local_ref then
        local res = self.__local_objects[value]
        -- res can be nil, it means the obj client referenced has been deleted
        if res == nil then
            error ('oid=' .. value)
        end
        return res
    end
end

function Connection:del_local_object(obj)
    for k, v in pairs(self) do
        if v == obj then
            self.__local_objects[k] = nil
        end
    end
end

return Connection

File: grpc_connection.py
# ! python2
# -*- coding: utf-8 -*-

import select
import socket
import pickle
import traceback
import types
import threading
import string
try:
    import queue as Queue
except:
    import Queue

import grpc_netref
import grpc_utils
from grpc_config import config


simple_types = frozenset(
    [type(None), int, long, bool, float, str, unicode, complex])


class Connection(object):
    ''' connection object based on socket, contains
    some functions that both server and client will use'''

    def __init__(self, buffer_size):
        self.__sock = None
        self.__buffer_size = buffer_size
        self.__seq_num = 1      # next request's sequence number
        self.__connected = False

        self.__local_objects = grpc_utils.CountDict()  # on server
        self.requests_cache = Queue.PriorityQueue()     # on server

        self.__netref_classes_cache = {}  # on client
        self.replies_cache = Queue.PriorityQueue()   # on client

    def __del__(self):
        self.shutdown()

    # network control
    # start
    # ...
    def accept(self, server_sock):
        client_sock, address = server_sock.accept()
        self.__sock = client_sock
        self.__connected = True
        self.__recv_thread = threading.Thread(target=self.__recv_forever)
        self.__recv_thread.start()
        return address

    def connect(self, server_address):
        if self.__connected:
            self.shutdown()
        while True:
            try:
                print 'connecting'
                self.__sock = socket.create_connection(
                    server_address, timeout=1)
                if self.__sock:
                    self.__connected = True
                    print 'connected'
                    self.__recv_thread = threading.Thread(
                        target=self.__recv_forever)
                    self.__recv_thread.start()
                    break
            except socket.timeout:
                print 'timeout'
            except:
                traceback.print_exc()
                raise
        return self.__connected

    def shutdown(self, flag=socket.SHUT_RDWR):
        if self.__connected:
            try:
                self.send_shutdown()
                self.__connected = False
                if hasattr(self, '_Connection__recv_thread'):
                    try:
                        self.__recv_thread.join()
                        print 'joined'
                    except:
                        pass
                self.__sock.shutdown(flag)
            except socket.error:
                pass
            except:
                traceback.print_exc()
            self.__local_objects.clear()
            self.__netref_classes_cache = {}
            self.__sock = None
            self.__seq_num = 1
    # ...
    # end
    # network control

    # send functions
    # start
    # ...
    def send_request(self, action_type, data):
        boxed_data = self.__box_request(data)
        res = self.__send(
            config.msg.request, self.__seq_num, action_type, boxed_data)
        if res > 0:
            self.__seq_num += 1
        else:
            self.shutdown()
        return res

    def send_reply(self, seq_num, action_type, data):
        boxed_data = self.__box_reply(data)
        res = self.__send(config.msg.reply, seq_num, action_type, boxed_data)
        if res > 0:
            return res
        self.shutdown()
        return -1

    def send_shutdown(self):
        try:
            return self.__send(config.msg.shutdown, 0, 0, 0)
        except socket.error:
            return -1

    def send_exception(self, seq_num, data):
        res = self.__send(config.msg.exception, seq_num, 0, data)
        if res > 0:
            return res
        self.shutdown()
        return -1

    def __send(self, msg_type, seq_num, action_type, boxed_data):
        if self.__connected is False:
            return -1
        try:
            pickled_data = pickle.dumps(
                (msg_type, seq_num, action_type, boxed_data))
        except:
            print 'data cannot be pickled'
            raise
        data_size = len(pickled_data)
        if data_size > 99999999:
            raise ValueError('data size is too large')
        try:
            self.__sock.sendall(str(data_size).zfill(8) + pickled_data)
        except socket.error:
            print 'socket error, shutdown'
            return -1
        return seq_num

    def __box_request(self, obj):
        if obj is NotImplemented:     # pickle cannot dump NotImplemented
            return config.label.notimplemented, None
        elif obj is Ellipsis:           # pickle cannot dump Ellipsis
            return config.label.ellipsis, None
        elif type(obj) is tuple:
            return config.label.tuple, tuple(
                self.__box_request(item) for item in obj)
        elif type(obj) is list:
            return config.label.list, tuple(
                self.__box_request(item) for item in obj)
        elif type(obj) is dict:
            return config.label.dict, tuple(
                self.__box_request(item) for item in obj.items())
        elif isinstance(obj, grpc_netref.NetRef) and obj.____conn__ is self:
            return config.label.local_ref, obj.____oid__
        else:
            return config.label.value, obj

    def __box_reply(self, obj):
        if type(obj) in simple_types:
            return config.label.value, obj
        elif obj is NotImplemented:     # pickle cannot dump NotImplemented
            return config.label.notimplemented, None
        elif obj is Ellipsis:           # pickle cannot dump Ellipsis
            return config.label.ellipsis, None
        elif type(obj) is tuple:
            return config.label.tuple, tuple(
                self.__box_reply(item) for item in obj)
        else:
            self.__local_objects.add(obj)
            try:
                cls = obj.__class__
            except:
                cls = type(obj)
            return config.label.remote_ref, (
                id(obj), cls.__name__, cls.__module__)
    # ...
    # end
    # send functions

    # recv functions
    # start
    # ...
    def recv(self, timeout=None):
        if self.__connected is False:
            return None
        data_size = 0
        recvd_size = 0

        # get data size
        ready = select.select([self.__sock], [], [], timeout)
        if ready[0]:
            data_size = self.__sock.recv(8)
            if not data_size:   # peer closed
                self.shutdown()
                return None
            else:
                data_size = int(data_size)
        else:
            return None

        # get data
        pickled_data = ""
        while data_size > recvd_size:
            ready = select.select([self.__sock], [], [], 1.0)
            if ready[0]:
                rest_data = self.__sock.recv(data_size - recvd_size)
                if not rest_data:   # peer closed
                    self.shutdown()
                    return None
                recvd_size += len(rest_data)
                pickled_data = "".join([pickled_data, rest_data])

        # unbox data
        msg_type, seq_num, action_type, data = pickle.loads(pickled_data)
        try:
            unboxed_data = None
            if msg_type == config.msg.request:
                unboxed_data = self.__unbox(data)
            elif msg_type == config.msg.reply:
                unboxed_data = self.__unbox(data)
            elif msg_type == config.msg.exception:
                unboxed_data = data
            elif msg_type == config.msg.shutdown:
                self.shutdown()
            return msg_type, seq_num, action_type, unboxed_data
        except KeyError as e:
            # send 'object has been del'
            # just on server
            error_msg = 'object oid=' + str(e) + ' has been deleted'
            self.send_exception(seq_num, error_msg)
        return None

    def __unbox(self, package):
        label, value = package
        if label == config.label.value:
            return value
        elif label == config.label.tuple:
            return tuple(self.__unbox(item) for item in value)
        elif label == config.label.list:
            return list(self.__unbox(item) for item in value)
        elif label == config.label.dict:
            return dict(self.__unbox(item) for item in value)
        elif label == config.label.notimplemented:
            return NotImplemented
        elif label == config.label.ellipsis:
            return Ellipsis
        elif label == config.label.local_ref:
            try:
                obj = self.__local_objects[value]
            except KeyError:
                raise KeyError(value)
            else:
                return obj
        elif label == config.label.remote_ref:
            oid, clsname, modname = value
            proxy = grpc_netref.NetRef(self, oid)
            return proxy
        else:
            raise ValueError("invalid label {}".format(label))

    def __recv_forever(self):
        while self.__connected:
            res = self.recv()
            # None means peer closed, or error occured
            # just continue
            if res is None:
                continue
            msg_type, seq_num, action_type, data = res
            if msg_type == config.msg.request:
                self.requests_cache.put((seq_num, res))
            elif msg_type == config.msg.reply:
                self.replies_cache.put((seq_num, res))
            elif msg_type == config.msg.exception:
                self.replies_cache.put((seq_num, res))

    # ...
    # end
    # recv functions

    # useful functions
    # start
    # ...
    def sync_request(self, action_type, data=None, need_reply=True):
        data = (need_reply, data)
        seq_num = self.send_request(action_type, data)
        if seq_num < 0 or not need_reply:
            return None
        while self.connected:
            if self.replies_cache.empty():
                continue
            (msg_type, recv_seq_num, action_type, recv_data) = (
                self.replies_cache.get()[1])
            if seq_num < recv_seq_num:
                self.replies_cache.put((
                    recv_seq_num,
                    (msg_type, recv_seq_num, action_type, recv_data)
                    ))
                continue
            if seq_num > recv_seq_num and msg_type != config.msg.exception:
                # recieved reply or exception of previous request
                # check if it is exception
                # if not, just ignore
                # if it is, raise it
                continue
            if msg_type == config.msg.reply:
                return recv_data
            if msg_type == config.msg.exception:
                if isinstance(recv_data, Exception):
                    raise recv_data
                elif isinstance(recv_data, types.StringType):
                    if string.find(recv_data, 'GrpcStopIteration') != -1:
                        raise StopIteration()
                    raise Exception(recv_data)
            return None

    @property
    def connected(self):
        return self.__connected

    def del_local_object(self, obj):
        del self.__local_objects[id(obj)]
    # ...
    # end
    # useful functions
